### ISSUE TITLE
Consider generating a meson toolchain file

### ISSUE THREAD

**Issue by eli-schwartz** [microsoft/vcpkg#36014]
https://learn.microsoft.com/en-us/vcpkg/users/buildsystems/cmake-integration

> When you set CMAKE_TOOLCHAIN_FILE to use the vcpkg toolchain (`<vcpkg-root>/scripts/buildsystems/vcpkg.cmake`), vcpkg takes advantage of the toolchain file mechanism to inject code to integrate with built-in CMake functions transparently to you.

Meson supports toolchain files as well: https://mesonbuild.com/Machine-files.html

Note in particular the `@DIRNAME@` constant which lets the toolchain find files distributed alongside the toolchain file itself.

It would be good to set up properties for the vcpkg root as well as automatically configure the pkg-config path and the cmake prefix path so that meson can automatically pick up dependencies.

This would help users who are currently not sure how to use vcpkg with meson:

https://mastodon.gamedev.place/@wolfpld/111562003666154458
https://github.com/wolfpld/tracy/issues/655#issuecomment-1875304238

(It does work as long as you hook it up. Vcpkg can get this right for everyone so that individual projects don't have to rediscover the same information.)

**Comment by Neumann-A** [comment]
a) missing a way to define options
b) missing a way to know the binary dir. (or is there a way to have @GLOBAL_BINARY_ROOT@)
c) missing a way to invoke vcpkg in manifest mode. 
d) missing a way to switch depending on configuration. 

So vcpkg can only set it up for release builds but that is not really enough on windows. 

> It does work as long as you hook it up. 

That is known since vcpkg can execute meson internally without a problem.

**Comment by eli-schwartz** [comment]
a) meson toolchain files can set any `[built-in options]`, equivalent to specifying `-D` options on the command line

b) I'm not sure I understand what getting the binary dir would be used for.

c) and d) is this just about using a scripting language to reduce a matrix?

d) specifically: I would assume you specifically want different toolchains for different target triplets / debug runtime

**Comment by Neumann-A** [comment]
a) vcpkg needs to have a way to define its own options (triplet selection logic)

b) In manifest mode the binary dir is the default installation location. In manifest mode only host/target installed trees exists so switching to a different combination would result in a removal of other installed combinations. 

c) no. Invoking vcpkg with the user supplied options automatically requires that. You are assuming the old classic mode where everything is preinstalled

d) not different toolchains but switching between debug and release configuration already requires setting up different pkg-config paths for meson. 

So basically you always need a "prepare meson machine file" step and "run vcpkg" step with the options provided later to meson to actually prepare a machine file which fulfills the needs of meson. Alternatively you can wrap your meson.build with another meson.build which exactly runs these steps. 

Basically the way meson is designed doesn't allow vcpkg to provide an easy and seamless working integration. It is all additional manual setup.

**Comment by eli-schwartz** [comment]
a) you don't need to define an option for this, since one already exists. Meson toolchain files already have a section for defining a host_machine triplet, in fact there isn't a -D option for it at all.

b) that sounds inconvenient given that the binary directory could get cleaned and rebuilt. I would assume it's a better idea to install to .vcpkg/ in the source directory instead. In such a case, vcpkg merely needs to describe that installation.

> So basically you always need a "prepare meson machine file" step and "run vcpkg" step with the options provided later to meson to actually prepare a machine file which fulfills the needs of meson. Alternatively you can wrap your meson.build with another meson.build which exactly runs these steps.

I don't understand why it's problematic to support people running a two step process just because you would prefer to run a one-step process. The current approach is a lot more than two steps -- improving that is a bad thing? Why?

And no, running a meson.build with another meson.build is not the answer. The meson.build file format is a DSL for describing projects and defining targets, it doesn't have an independent script mode for using it as a general purpose scripting language. You can, of course, wrap your meson.build with a meson-vcpkg.py script instead. But it would help out vcpkg users if they did not have to because vcpkg wrote out a toolchain file for them *automatically*.

**Comment by jimwang118** [comment]
Thanks for posting this issue. Please reopen this issue if this is still a problem for you.

**Comment by eli-schwartz** [comment]
> Thanks for posting this issue. Please reopen this issue if this is still a problem for you.

I do not possess the power to *reopen this issue*. I don't have triage permissions here.

The feature request was not as far as I can tell implemented, thus it is still a problem.

**Comment by ethindp** [comment]
It would be nice if this got actually done. I would love to use vcpkg with meson.
